### PR TITLE
Update to 6.x

### PR DESCRIPTION
**Pull request by jschueller**
resurrecting https://github.com/conda-forge/qt-main-feedstock/pull/14 against the new qt6 branch

### CODE REVIEW

**Comment by jschueller** [comment]
@conda-forge-admin  please rerender

**Comment by hoechenberger** [comment]
Great work!!

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by ccordoba12** [comment]
Thanks @jschueller for resuming your work on this! A quick review according to your comment above:

> most previous patches are relevant to the old buildsystem, for example the new-dtags patch that is now a cmake option

Before proceeding, please rework your current git history and document in it every patch that needs to be dropped because it was necessary only for the old build system.

> some patches have been applied upstream (eg 0030-include_uitility_mbgl_unique_any.patch)

Also document this by adding a commit per dropped patch.

> disabled osx/arm64 for now, requires to have the native+cross compiler toolchains at the same time to build first a native version of some qt tools then build for host with -DQT_HOST_DIR=BUILD_PREFIX, so disabled the arm patches too

Ok, I understand that. Please leave a single commit when doing it and mention why that's necessary for now in the commit message.

> for the tests we can rely on the qt provided examples instead of the old recipe /test folder

As far as I can tell, you're only building a Qt example code, but the previous tests also ran their example code. Is it possible to do that with the new test?

**Comment by ccordoba12** [comment]
In essence, we'd like to see a more detailed git history that future maintainers can use as a reference to understand why any change to the current recipes was necessary when using `git blame`.

Please take a look at https://github.com/conda-forge/qt-feedstock/pull/206/commits for an example. I'm not expecting such a detailed history in this case because building Qt6 is totally different, but you can get an idea from it on how to structure your commits.

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@ccordoba12 better now ?

**Comment by Tobias-Fischer** [comment]
This is really cool, thanks @jschueller!

Can you please help me understand what the issue on osx-arm64 is? We have several feedstocks where we build some tools first, and then use them for the "actual" build (apologies for not using proper terminology; I'm not too deep in the material). See e.g. https://github.com/conda-forge/gazebo-feedstock/blob/main/recipe/build.sh - note that in this case the `host` deps also need to be added to `build`: https://github.com/conda-forge/gazebo-feedstock/blob/main/recipe/meta.yaml

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by Tobias-Fischer** [comment]
Looks good so far - now we'll need to add the host deps to build as in https://github.com/conda-forge/gazebo-feedstock/blob/main/recipe/meta.yaml

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
thanks @Tobias-Fischer

**Comment by Tobias-Fischer** [comment]
Great work, thanks to you @jschueller! If someone gives green light to the changes made in the PR, I can give the linux-aarch64 build a go.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -128,6 +115,23 @@ requirements:
     # cannot be executed on x86_64 to locate libclang successfully
     - llvmdev                           # [arm64]
     - clangdev                          # [arm64]
+    # native libs
+    - icu                               # [arm64]
```
Comment (isuruf) [inline recipe/meta.yaml:119]:
```suggestion
    - icu                               # [build_platform != target_platform]
```

Comment (jschueller) [reply recipe/meta.yaml:119]: ok

File: recipe/meta.yaml [hunk 2]
```diff
@@ -128,6 +115,23 @@ requirements:
     # cannot be executed on x86_64 to locate libclang successfully
     - llvmdev                           # [arm64]
     - clangdev                          # [arm64]
```
Comment (isuruf) [inline recipe/meta.yaml:117]: Same for the three lines above

**Comment by isuruf** [comment]
@conda-forge-admin, rerender

**Comment by jschueller** [comment]
@isuruf on aarch64 I get an incomprehensible segfault in QFileSystemEngine::fillMetaData
(https://github.com/qt/qtbase/blob/v6.3.1/src/corelib/io/qfilesystemengine_unix.cpp#L427) for the native (x86_64) build
do you have any clues ?
```
2022-09-05T05:59:39.3268941Z Program received signal SIGABRT, Aborted.
2022-09-05T05:59:39.3269424Z 0x00007f94389d2387 in raise () from /lib64/libc.so.6
2022-09-05T05:59:39.3305950Z #0  0x00007f94389d2387 in raise () from /lib64/libc.so.6
2022-09-05T05:59:39.3306609Z #1  0x00007f94389d3a78 in abort () from /lib64/libc.so.6
2022-09-05T05:59:39.3354732Z #2  0x00007f9438a14f67 in __libc_message () from /lib64/libc.so.6
2022-09-05T05:59:39.3355618Z #3  0x00007f9438ab47a7 in __fortify_fail () from /lib64/libc.so.6
2022-09-05T05:59:39.3889355Z #4  0x00007f9438ab4762 in __stack_chk_fail () from /lib64/libc.so.6
2022-09-05T05:59:39.4766316Z #5  0x0000559e5ccae720 in QFileSystemEngine::fillMetaData (fd=<optimized out>, data=...) at /usr/local/src/conda/qt-main-6.3.1/qtbase/src/corelib/io/qfilesystemengine_unix.cpp:427
2022-09-05T05:59:39.5142218Z #6  0x0000559e5ccb23ee in QFSFileEnginePrivate::nativeOpenImpl (this=0x559e5e3cff30, openMode=..., mode=438) at /usr/local/src/conda/qt-main-6.3.1/qtbase/src/corelib/io/qfsfileengine_unix.cpp:144
2022-09-05T05:59:39.6590735Z #7  0x0000559e5cc7ac9f in QFSFileEngine::open (this=0x559e5e3cff10, openMode=..., permissions=...) at /usr/local/src/conda/qt-main-6.3.1/qtbase/src/corelib/io/qfsfileengine.cpp:254
2022-09-05T05:59:39.6593133Z #8  0x0000559e5cc75f67 in QFile::open (this=this@entry=0x7ffe522e8df0, mode=..., mode@entry=...) at /usr/local/src/conda/qt-main-6.3.1/build_host/qtbase/include/QtCore/../../../../qtbase/src/corelib/global/qflags.h:70
```

**Comment by Tobias-Fischer** [comment]
Maybe adding `xorg-libxfixes  # [linux]` to the host deps will fix the issue that GL.h isn't found?

**Comment by Tobias-Fischer** [comment]
Error remains: `$SRC_DIR/build_host/qtbase/include/QtGui/../../../../qtbase/src/gui/opengl/qopengl.h:141:13: fatal error: GL/gl.h: No such file or directory`

I think it is because it finds the system OpenGL instead of the conda one:
```
2022-09-05T22:46:58.7158756Z // Path to a file.
2022-09-05T22:46:58.7159267Z OPENGL_EGL_INCLUDE_DIR:PATH=/usr/include
2022-09-05T22:46:58.7159777Z 
2022-09-05T22:46:58.7160193Z // Path to a file.
2022-09-05T22:46:58.7161144Z OPENGL_GLX_INCLUDE_DIR:PATH=/usr/include
2022-09-05T22:46:58.7161422Z 
2022-09-05T22:46:58.7161724Z // Path to a file.
2022-09-05T22:46:58.7162081Z OPENGL_INCLUDE_DIR:PATH=/usr/include
2022-09-05T22:46:58.7162317Z 
2022-09-05T22:46:58.7162592Z // Path to a library.
2022-09-05T22:46:58.7163142Z OPENGL_egl_LIBRARY:FILEPATH=/usr/lib64/libEGL.so
2022-09-05T22:46:58.7163375Z 
2022-09-05T22:46:58.7163658Z // Path to a library.
2022-09-05T22:46:58.7164325Z OPENGL_glu_LIBRARY:FILEPATH=OPENGL_glu_LIBRARY-NOTFOUND
2022-09-05T22:46:58.7164635Z 
2022-09-05T22:46:58.7164916Z // Path to a library.
2022-09-05T22:46:58.7165314Z OPENGL_glx_LIBRARY:FILEPATH=/usr/lib64/libGLX.so
2022-09-05T22:46:58.7165547Z 
2022-09-05T22:46:58.7165813Z // Path to a library.
2022-09-05T22:46:58.7188181Z OPENGL_opengl_LIBRARY:FILEPATH=/usr/lib64/libOpenGL.so
```

**Comment by jschueller** [comment]
@Tobias-Fischer seems xorg-libxfixes is not enough to bring gl.h

**Comment by Tobias-Fischer** [comment]
What happens if you don't hide the host libs? I haven't seen this before and I'm not sure whether this is the right way to go.

**Comment by jschueller** [comment]
If I dont hist the host libs I get conflicts from cmake:
```
runtime library [libEGL.so.1] in /usr/lib64 may be hidden by files in:
2022-09-05T05:53:32.9941419Z       /home/conda/feedstock_root/build_artifacts/qt-main_1662355988577/_build_env/aarch64-conda-linux-gnu/sysroot/usr/lib
```
then moc crashes the native build, I'm guessing it somehow runs but host libs are interfering

**Comment by Tobias-Fischer** [comment]
Out of disk space :(.

Maybe `FEATURE_optimize_size` helps?

**Comment by isuruf** [comment]
> Maybe FEATURE_optimize_size helps?

Why? It optimizes the binary, not the build process.

**Comment by Tobias-Fischer** [comment]
Ok, that makes sense. In this case can we disable some features for aarch?

**Comment by ccordoba12** [comment]
> Ok, that makes sense. In this case can we disable some features for aarch?

I doubt aarch could pass on CIs because it's emulated inside Qemu (if I'm not mistaken), so it's very slow.

**Comment by Tobias-Fischer** [comment]
No, it's cross-compiled. It's getting to `[5475/6044] Running moc --collect-json for target Designer` so reasonably close to the end of the build, with another hour to spare in CI.

File: recipe/bld.bat [hunk 3]
```diff
@@ -1,128 +1,16 @@
...
-if exist config.cache del config.cache
+mkdir build && cd build
+cmake -LAH -G "Ninja" ^
+    -DCMAKE_BUILD_TYPE=Release ^
+    -DCMAKE_PREFIX_PATH="%LIBRARY_PREFIX%" ^
+    -DCMAKE_INSTALL_PREFIX="%LIBRARY_PREFIX%" ^
+    -DINSTALL_INCLUDEDIR=include/qt6 ^
+    -DINSTALL_MKSPECSDIR=lib/qt6/mkspecs ^
+    -DINSTALL_DATADIR=share/qt6 ^
+    -DQT_BUILD_SUBMODULES="qt3d;qtbase;qtcharts;qtdatavis3d;qtdeclarative;qtimageformats;qtmultimedia;qtnetworkauth;qtremoteobjects;qtshadertools;qtsvg;qttools;qttranslations" ^
```
Comment (ccordoba12) [inline recipe/bld.bat:10]: Could you move this list to a batch variable formatted vertically (i.e. one Qt library per line) and then use it here?

I think that way it'd be quite easy to see what libraries we're building and make sense of the diff when adding new ones.

File: recipe/build.sh [hunk 4]
```diff
@@ -1,308 +1,74 @@
...
+  -DINSTALL_INCLUDEDIR=include/qt6 \
+  -DINSTALL_MKSPECSDIR=lib/qt6/mkspecs \
+  -DINSTALL_EXAMPLESDIR=share/doc/qt6/examples \
+  -DFEATURE_system_sqlite=ON \
+  -DFEATURE_framework=OFF \
+  -DFEATURE_linux_v4l=OFF \
+  -DFEATURE_gssapi=OFF \
+  -DFEATURE_enable_new_dtags=OFF \
+  -DFEATURE_gstreamer_gl=OFF \
+  -DQT_BUILD_SUBMODULES="qt3d;qtbase;qtcharts;qtdatavis3d;qtdeclarative;qtimageformats;qtmultimedia;qtnetworkauth;qtremoteobjects;qtshadertools;qtsvg;qttools;qttranslations" \
```
Comment (ccordoba12) [inline recipe/build.sh:72]: Could you do the same here?

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,51 +1,37 @@
...
 
 package:
   name: {{ name|lower }}
   version: {{ version }}
 
 source:
-  - url: https://download.qt.io/official_releases/qt/{{ version.rpartition('.')[0] }}/{{ version }}/single/qt-everywhere-opensource-src-{{ version }}.tar.xz
-    sha256: b7412734698a87f4a0ae20751bab32b1b07fdc351476ad8e35328dbe10efdedb
+  - url: https://download.qt.io/official_releases/qt/{{ version.rpartition('.')[0] }}/{{ version }}/single/qt-everywhere-src-{{ version }}.tar.xz
+    md5: 957a304773b281a4584f4c0254773456
```
Comment (ccordoba12) [inline recipe/meta.yaml:10]: Is it possible to use sha256 here?

**Review by ccordoba12 — COMMENTED** [review]
Some small suggestions from my side.

File: recipe/meta.yaml (hunk 5 above)
Comment (jschueller) [reply recipe/meta.yaml:10]: qt only provides md5s, so its easier:
https://download.qt.io/official_releases/qt/6.3/6.3.1/single/

**Comment by jschueller** [comment]
> Ok, that makes sense. In this case can we disable some features for aarch?

I'll try to disable pch before disabling libs

File: recipe/bld.bat (hunk 3 above)
Comment (jschueller) [reply recipe/bld.bat:10]: yes

File: recipe/build.sh (hunk 4 above)
Comment (jschueller) [reply recipe/build.sh]: done

**Comment by Tobias-Fischer** [comment]
Awesome, the linux-aarch64 built successfully now. However osx-arm64 is timing out - have you changed anything there, or was it just a randomly slow build?

**Comment by jschueller** [comment]
I think it was random

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
It should be all ok now

**Comment by jschueller** [comment]
Everything is green!

**Comment by jschueller** [comment]
Sounds like a plan, I'll be happy to address any additional comments

**Comment by Tobias-Fischer** [comment]
Awesome work indeed, many thanks!

I've got some questions: 1) can qt5 and qt6 be installed in parallel? If so, we should allow this by e.g. renaming the package to qt-main6 or something like that? 2) is the plan to also build qt-webengine6? 3) how do you envision the migration to qt6?

**Comment by jschueller** [comment]
This is possible with versioned install dirs and versioned symlinks, I just found out that qt has thought about that, not sure how it will play out on win32 though, see last commit.
Indeed the usual way to do that is to have versioned packages too.
For webengine I dont know if it will also require a two-step build stage for more host tools for cross-compilation, but native builds should at least be doable.
But most packages cannot use qt6 yet, eg paraview is years away.

**Comment by ccordoba12** [comment]
> This is possible with versioned install dirs and versioned symlinks, I just found out that qt has thought about that, not sure how it will play out on win32 though, see last commit.

I don't think that's enough for conda @jschueller. As @Tobias-Fischer mentioned, unless the package is also renamed to `qt-main6` (or better yet `qt6-main`), conda would simply remove the previous `qt-main` version and install the new one.

Besides, there's the issue of `qt.conf`: I think we'd need two `qt.conf`'s, one for version 5 and another for 6 so their packages can be relocated correctly. I don't know if that could introduce too many issues for downstream feedstocks that are not ready to support Qt 6.

Finally, this split of Qt5/6 would need to be propagated to PyQt5/6 and PySide2/6, which could be a nightmare to support down the road.

**Comment by jschueller** [comment]
conf files are already versioned  (altough I dont see any qt.conf file)
so I think its just a matter of renaming then

**Comment by looooo** [comment]
why installing qt5 / qt6 into one environment? I thought this is the great advantage of the environments: there is no need for having multiple versions of one library in one env. Having multiple versions of one library in one environment is always a source of issues, imo.

**Comment by hoechenberger** [comment]
I also don't see why renaming should be necessary. If any downstream packages don't support Qt6 yet, well, then their dependency specs must be fixed to express that.

**Comment by hmaarrfk** [comment]
Why are we making a qt6 branch instead of having a "maintenance" qt5 branch?

**Comment by jschueller** [comment]
qt5 remains the most frequent branch used by downstream packages by far, so it makes sense to keep qt5 as master, at least for now

**Comment by hmaarrfk** [comment]
I think you will have trouble doing a name swap on git in the future. but as you wish.

**Comment by ccordoba12** [comment]
> Why are we making a qt6 branch instead of having a "maintenance" qt5 branch?

That was my decision, following a similar reasoning as the one expressed by @jschueller. I mentioned it in #14 and no one said anything against it at the time. But we could revise it if you think that was not the best call.

> I also don't see why renaming should be necessary.

I also agree with this because it'd make maintenance way simpler than if we create a new set of packages for Qt6 (qt6-main, qt6-webengine, pyqt6, pyside6, etc). Not to mention the confusion this could introduce to users.

**Comment by Tobias-Fischer** [comment]
So if we don't allow installation of qt5 and qt6 side-by-side, and are not careful in downstream feedstocks (where some maintainers might opt to manually migrate to qt6 without also still building for qt5), won't we fairly soon run into the situation where packages become incompatible with each other? We would need to maintain two branches of qt5/6-webengine etc. anyhow, regardless of the renaming.

Note that I don't have particularly strong feelings about this, I am just curious how we envision the transition to qt6 :).

**Comment by hmaarrfk** [comment]
lets discuss the migration plan https://github.com/conda-forge/qt-main-feedstock/issues/49

**Comment by Tobias-Fischer** [comment]
That looks great, thanks @jschueller! Should we also check in the tests whether a simple qt5 executable can still be built? In this way we'd be very certain that qt5 and qt6 can not only installed but also both be used at the same time.

File: recipe/conda_build_config.yaml [hunk 6]
```diff
@@ -1,4 +1,18 @@
...
-  - '10.13'                # [osx and x86_64]
-MACOSX_SDK_VERSION:        # [osx and x86_64]
   - '10.14'                # [osx and x86_64]
+MACOSX_SDK_VERSION:        # [osx and x86_64]
+  - '10.15'                # [osx and x86_64]
+
+c_compiler:                # [win]
+  - vs2019                 # [win]
+  - vs2019                 # [win]
+  - vs2019                 # [win]
```
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml:9]: I believe this is now the new default.

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

File: recipe/conda_build_config.yaml (hunk 6 above)
Comment (jschueller) [reply recipe/conda_build_config.yaml]: indeed, thanks

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

File: recipe/meta.yaml [hunk 7]
```diff
@@ -165,23 +181,39 @@ requirements:
...
+    - xcb-util                           # [linux]
+    - xcb-util-wm                        # [linux]
+    - xcb-util-keysyms                   # [linux]
+    - xcb-util-image                     # [linux]
+    - xcb-util-renderutil                # [linux]
     - {{ pin_compatible("libclang") }}
   run_constrained:
     - __osx >={{ MACOSX_DEPLOYMENT_TARGET|default("10.14") }}  # [osx and x86_64]
     - qt {{ version }}
+    - qt-main >=999  # [win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:205]:
```suggestion
    # windows symbols clash with qt5. See upstream issue:
    # LINK TO UPSTREAM 
    - qt-main >=999  # [win]
```

Comment (jschueller) [reply recipe/meta.yaml:207]: ok

**Comment by jschueller** [comment]
mentioned upstream issue:
https://bugreports.qt.io/browse/QTBUG-107009

File: conda-forge.yml [hunk 8]
```diff
@@ -4,6 +4,7 @@ azure:
       CONDA_BLD_PATH: C:\bld\
 build_platform:
   osx_arm64: osx_64
+  linux_aarch64: linux_64
```
Comment (hmaarrfk) [inline conda-forge.yml:7]: Is cross compiling now supported! So much easier to build!

Comment (Tobias-Fischer) [reply conda-forge.yml:7]: On that note, shall we try cross-compiling for ppc?

File: recipe/conda_build_config.yaml [hunk 9]
```diff
@@ -1,4 +1,9 @@
 MACOSX_DEPLOYMENT_TARGET:  # [osx and x86_64]
-  - '10.13'                # [osx and x86_64]
-MACOSX_SDK_VERSION:        # [osx and x86_64]
   - '10.14'                # [osx and x86_64]
+MACOSX_SDK_VERSION:        # [osx and x86_64]
+  - '10.15'                # [osx and x86_64]
+
+c_compiler_version:        # [osx]
```
Comment (Tobias-Fischer) [inline recipe/conda_build_config.yaml:6]: Could you please add a comment why this is necessary?

**Comment by Tobias-Fischer** [comment]
Overall this looks great to me, thanks again @jschueller!

A few comments:
  - Do you want to add yourself to the maintainers?
  - Can we delete the [qt.conf](https://github.com/jschueller/qt-main-feedstock/blob/qt6x/recipe/qt.conf) and [write_qtconf.bat](https://github.com/jschueller/qt-main-feedstock/blob/qt6x/recipe/write_qtconf.bat) and [tst_compiler.cpp](https://github.com/jschueller/qt-main-feedstock/blob/qt6x/recipe/tst_compiler.cpp) files? I don't think they are still used, are they?

File: conda-forge.yml (hunk 8 above)
Comment (hmaarrfk) [reply conda-forge.yml:7]: probably after this PR is in.

**Comment by jschueller** [comment]
@Tobias-Fischer done

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by Tobias-Fischer** [comment]
Awesome, green light from me.

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
@conda-forge-admin please rerender

**Comment by jschueller** [comment]
Ready, updated to 6.4.0

**Comment by jschueller** [comment]
update: worked around the qt-main 5.x conflict on win32

**Comment by jschueller** [comment]
update: the new qt6-main package does not conflict with qt-main (5.x), so this should let us moving forward with the 6.x version without worrying of breaking 5.x-only packages (migration strategy can still be discussed in #49)
also, this is blocking me from submitting pyside6

**Comment by hoechenberger** [comment]
Amazing!

**Pull request merged**